src/Database/LSMTree/Internal/MergeSchedule.hs

```diff
@@ -168,19 +172,24 @@ data MergingRun m h =
 -- TODO: this should also represent ongoing merges once we implement scheduling.
 newtype MergingRunState m h = CompletedMerge (Run m h)
 
--- | Return all runs in the levels, ordered from newest to oldest
-runsInLevels :: Levels m h -> V.Vector (Run m h)
-runsInLevels levels = flip V.concatMap levels $ \(Level mr rs) ->
+{-# SPECIALISE forRunM_ :: Levels IO h -> (Run IO h -> IO ()) -> IO () #-}
+forRunM_ :: PrimMonad m => Levels m h -> (Run m h -> m ()) -> m ()
+forRunM_ lvls k = V.forM_ lvls $ \(Level mr rs) -> do
     case mr of
-      SingleRun r                   -> r `V.cons` rs
-      MergingRun (CompletedMerge r) -> r `V.cons` rs
-
-{-# SPECIALISE closeLevels :: Levels IO (Handle h) -> IO () #-}
-closeLevels ::
-     m ~ IO -- TODO: replace by @io-classes@ constraints for IO simulation.
-  => Levels m (Handle h)
-  -> m ()
-closeLevels levels = V.mapM_ Run.removeReference (runsInLevels levels)
+      SingleRun r    -> k r
+      MergingRun var -> readMutVar var >>= \case
+        CompletedMerge r -> k r
+    V.mapM_ k rs
+
+{-# SPECIALISE forRunM :: Levels IO h -> (Run IO h -> IO a) -> IO (V.Vector a) #-}
+forRunM :: PrimMonad m => Levels m h -> (Run m h -> m a) -> m (V.Vector a)
```

Comment:
It's not clear that we need this, and its implementation isn't terribly efficient (with lots of intermediate vectors and a concat map at the end).

What is used is a function to read and return a vector of all runs in all levels (currently used by `mkLevelsCache`). Such a function could be made efficient if necessary.

Reply:
I've replaced intermediate vectors with intermediate lists. It is probably still not super performant, but I've left a TODO that mentions that we might not need this function in the end.